--- ../alsa-kernel/core/hrtimer.c	2015-07-24 20:09:03.433652027 +0200
+++ hrtimer.c	2015-07-27 12:54:09.682380313 +0200
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /*
  * ALSA timer back-end using hrtimer
  * Copyright (C) 2008 Takashi Iwai
@@ -67,6 +68,11 @@
 		return -ENOMEM;
 	hrtimer_init(&stime->hrt, CLOCK_MONOTONIC, HRTIMER_MODE_REL);
 	stime->timer = t;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 28)
+	stime->hrt.cb_mode = HRTIMER_CB_IRQSAFE;
+#elif LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 29) && defined(HRTIMER_STATE_PENDING)
+	stime->hrt.cb_mode = HRTIMER_CB_IRQSAFE_UNLOCKED;
+#endif
 	stime->hrt.function = snd_hrtimer_callback;
 	atomic_set(&stime->running, 0);
 	t->private_data = stime;
@@ -123,7 +129,13 @@
 	struct snd_timer *timer;
 	int err;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 2, 0)
 	resolution = hrtimer_resolution;
+#else
+	struct timespec tp;
+	hrtimer_get_res(CLOCK_MONOTONIC, &tp);
+	resolution = tp.tv_nsec;
+#endif
 
 	/* Create a new timer and set up the fields */
 	err = snd_timer_global_new("hrtimer", SNDRV_TIMER_GLOBAL_HRTIMER,
