--- ../alsa-kernel/core/sound.c	2015-04-27 12:25:57.385619344 +0200
+++ sound.c	2015-04-27 17:42:20.772089166 +0200
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /*
  *  Advanced Linux Sound Architecture
  *  Copyright (c) by Jaroslav Kysela <perex@perex.cz>
@@ -149,6 +150,15 @@
 #define autoload_device(minor)	NULL
 #endif /* CONFIG_MODULES */
 
+#ifndef replace_fops
+#define replace_fops(f, fops) \
+	do {	\
+		struct file *__file = (f); \
+		fops_put(__file->f_op); \
+		BUG_ON(!(__file->f_op = (fops))); \
+	} while(0)
+#endif
+
 static int snd_open(struct inode *inode, struct file *file)
 {
 	unsigned int minor = iminor(inode);
@@ -397,6 +407,12 @@
  *  INIT PART
  */
 
+#ifdef CONFIG_SND_DEBUG_MEMORY
+extern void snd_memory_done(void);
+#else
+#define snd_memory_done()
+#endif
+
 static int __init alsa_sound_init(void)
 {
 	snd_major = major;
@@ -407,6 +423,7 @@
 	}
 	if (snd_info_init() < 0) {
 		unregister_chrdev(major, "alsa");
+		snd_memory_done();
 		return -ENOMEM;
 	}
 #ifndef MODULE
@@ -419,7 +436,10 @@
 {
 	snd_info_done();
 	unregister_chrdev(major, "alsa");
+	snd_memory_done();
 }
 
 subsys_initcall(alsa_sound_init);
 module_exit(alsa_sound_exit);
+
+#include "sound.inc"
