--- ../../alsa-kernel/pci/echoaudio/indigodjx.c
+++ indigodjx.c
@@ -1,3 +1,8 @@
+#include "adriver.h"
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 5, 0)
+#define pci_device(chip) pci_name(chip->pci)
+#endif
 /*
  *  ALSA driver for Echoaudio soundcards.
  *  Copyright (C) 2009 Giuliano Pochini <pochini@shiny.it>
@@ -105,3 +110,5 @@
 #include "indigo_express_dsp.c"
 #include "echoaudio_dsp.c"
 #include "echoaudio.c"
+
+EXPORT_NO_SYMBOLS;
