--- ../alsa-kernel/core/pcm.c	2012-10-17 09:25:29.000000000 +0200
+++ pcm.c	2012-10-25 01:08:45.000000000 +0200
@@ -1,3 +1,23 @@
+#include "adriver.h"
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 4, 0)
+#include <linux/compat.h>
+static inline int get_compat_timespec(struct timespec *ts,
+				      const struct compat_timespec __user *cts)
+{
+	return (!access_ok(VERIFY_READ, cts, sizeof(*cts)) ||
+			__get_user(ts->tv_sec, &cts->tv_sec) ||
+			__get_user(ts->tv_nsec, &cts->tv_nsec)) ? -EFAULT : 0;
+}
+
+static inline int put_compat_timespec(const struct timespec *ts,
+				      struct compat_timespec __user *cts)
+{
+	return (!access_ok(VERIFY_WRITE, cts, sizeof(*cts)) ||
+			__put_user(ts->tv_sec, &cts->tv_sec) ||
+			__put_user(ts->tv_nsec, &cts->tv_nsec)) ? -EFAULT : 0;
+}
+#endif
+
 /*
  *  Digital Audio (PCM) abstract layer
  *  Copyright (c) by Jaroslav Kysela <perex@perex.cz>
@@ -991,8 +1011,12 @@
 	substream->pstr->substream_opened--;
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 20)
 static ssize_t show_pcm_class(struct device *dev,
 			      struct device_attribute *attr, char *buf)
+#else
+static ssize_t show_pcm_class(struct class_device *class_device, char *buf)
+#endif
 {
 	struct snd_pcm *pcm;
 	const char *str;
@@ -1003,7 +1027,11 @@
 		[SNDRV_PCM_CLASS_DIGITIZER] = "digitizer",
 	};
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 20)
 	if (! (pcm = dev_get_drvdata(dev)) ||
+#else
+	if (! (pcm = class_get_devdata(class_device)) ||
+#endif
 	    pcm->dev_class > SNDRV_PCM_CLASS_LAST)
 		str = "none";
 	else
@@ -1011,7 +1039,11 @@
         return snprintf(buf, PAGE_SIZE, "%s\n", str);
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 20)
 static struct device_attribute pcm_attrs =
+#else
+static struct class_device_attribute pcm_attrs =
+#endif
 	__ATTR(pcm_class, S_IRUGO, show_pcm_class, NULL);
 
 static int snd_pcm_dev_register(struct snd_device *device)
@@ -1063,7 +1095,7 @@
 			return err;
 		}
 		snd_add_device_sysfs_file(devtype, pcm->card, pcm->device,
-					  &pcm_attrs);
+					  (struct device_attribute *)&pcm_attrs);
 		for (substream = pcm->streams[cidx].substream; substream; substream = substream->next)
 			snd_pcm_timer_init(substream);
 	}
