--- ../../alsa-kernel/include/control.h	2015-02-03 17:57:48.469745052 +0100
+++ control.h	2015-02-04 16:07:37.477453819 +0100
@@ -61,6 +61,9 @@
 
 struct snd_kcontrol_volatile {
 	struct snd_ctl_file *owner;	/* locked */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+	pid_t owner_pid;	/* just for kABI compatibility */
+#endif
 	unsigned int access;	/* access rights */
 };
 
@@ -102,13 +105,28 @@
 struct snd_ctl_file {
 	struct list_head list;		/* list of all control files */
 	struct snd_card *card;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#ifdef __GENKSYMS__
+	pid_t pid;
+#else
+	pid_t kabi_compat_pid;		/* just for kABI compatibility */
+#endif
+#else
 	struct pid *pid;
+#endif
 	int preferred_subdevice[SND_CTL_SUBDEV_ITEMS];
 	wait_queue_head_t change_sleep;
 	spinlock_t read_lock;
 	struct fasync_struct *fasync;
 	int subscribed;			/* read interface is activated */
 	struct list_head events;	/* waiting events for read */
+
+	/* appended below for ABI compatibility */
+#ifndef __GENKSYMS__
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+	struct pid *pid;
+#endif
+#endif
 };
 
 #define snd_ctl_file(n) list_entry(n, struct snd_ctl_file, list)
