--- ../../alsa-kernel/include/core.h	2014-03-01 11:24:19.732897531 +0100
+++ core.h	2014-03-01 11:30:39.808708785 +0100
@@ -100,8 +100,16 @@
 	char shortname[32];		/* short name of this soundcard */
 	char longname[80];		/* name of this soundcard */
 	char mixername[80];		/* mixer name */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 27)
+#ifdef __GENKSYMS__
+	char components[80];		/* just for ABI compatibility */
+#else
+	char old_components[80];	/* just for ABI compatibility */
+#endif
+#else
 	char components[128];		/* card components delimited with
 								space */
+#endif
 	struct module *module;		/* top-level module */
 
 	void *private_data;		/* private data for soundcard */
@@ -121,7 +129,15 @@
 	struct snd_info_entry *proc_id;	/* the card id */
 	struct proc_dir_entry *proc_root_link;	/* number link to real id */
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 30)
+#ifdef __GENKSYMS__
+	struct snd_monitor_file *files;	/* just for ABI compatibility */
+#else
+	void *old_monitor_file_next;	/* just for ABI compatibility */
+#endif
+#else
 	struct list_head files_list;	/* all files associated to this card */
+#endif
 	struct snd_shutdown_f_ops *s_f_ops; /* file operations in the shutdown
 								state */
 	spinlock_t files_lock;		/* lock the files for this card */
@@ -141,6 +157,17 @@
 	struct snd_mixer_oss *mixer_oss;
 	int mixer_oss_change_count;
 #endif
+
+	/* added in the tail */
+#ifndef __GENKSYMS__
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 27)
+	char components[128];		/* card components delimited with
+								space */
+#endif
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 30)
+	struct list_head files_list;	/* all files associated to this card */
+#endif
+#endif /* !__GENKSYMS__ */
 };
 
 #define dev_to_snd_card(p)	container_of(p, struct snd_card, card_dev)
@@ -301,7 +328,11 @@
 int snd_card_file_remove(struct snd_card *card, struct file *file);
 #define snd_card_unref(card)	put_device(&(card)->card_dev)
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 0)
+#define snd_card_set_dev(card,dev) /* no struct device */
+#else
 #define snd_card_set_dev(card, devptr) ((card)->dev = (devptr))
+#endif
 
 /* device.c */
 
@@ -428,10 +459,12 @@
 
 /* for easier backward-porting */
 #if defined(CONFIG_GAMEPORT) || defined(CONFIG_GAMEPORT_MODULE)
+#ifndef gameport_set_dev_parent
 #define gameport_set_dev_parent(gp,xdev) ((gp)->dev.parent = (xdev))
 #define gameport_set_port_data(gp,r) ((gp)->port_data = (r))
 #define gameport_get_port_data(gp) (gp)->port_data
 #endif
+#endif
 
 /* PCI quirk list helper */
 struct snd_pci_quirk {
@@ -444,6 +477,12 @@
 #endif
 };
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 30)
+/* rename for ABI compatibility */
+#define snd_pci_quirk		snd_pci_new_quirk
+#define snd_pci_quirk_lookup	snd_pci_new_quirk_lookup
+#endif
+
 #define _SND_PCI_QUIRK_ID_MASK(vend, mask, dev)	\
 	.subvendor = (vend), .subdevice = (dev), .subdevice_mask = (mask)
 #define _SND_PCI_QUIRK_ID(vend, dev) \
