--- ../../alsa-kernel/include/pcm.h	2015-09-13 12:03:42.003506463 +0200
+++ pcm.h	2015-09-24 14:52:36.974428470 +0200
@@ -444,7 +444,15 @@
 	int number;
 	char name[32];			/* substream name */
 	int stream;			/* stream (direction) */
+#ifdef CONFIG_SND_HAVE_PM_QOS_REQUEST_STATIC   
+	struct pm_qos_request_list latency_pm_qos_req; /* pm_qos request */
+#elif defined(CONFIG_SND_HAVE_PM_QOS_REQUEST_LIST)
+	struct pm_qos_request_list *latency_pm_qos_req; /* pm_qos request */
+#elif defined(CONFIG_SND_HAVE_PM_QOS_REQUEST)
 	struct pm_qos_request latency_pm_qos_req; /* pm_qos request */
+#else
+	char latency_id[20];
+#endif        
 	size_t buffer_bytes_max;	/* limit ring buffer size */
 	struct snd_dma_buffer dma_buffer;
 	size_t dma_max;
