--- ../alsa-kernel/firewire/cmp.c	2013-10-22 10:03:16.835007244 +0200
+++ cmp.c	2013-10-25 14:27:13.144974424 +0200
@@ -1,3 +1,11 @@
+#define __NO_VERSION__
+#include "adriver.h"
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 5, 0)
+const char *fw_rcode_string(int rcode);
+#endif
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(2, 6, 39)
+#include "iso-resources-old.h"
+#endif
 /*
  * Connection Management Procedures (IEC 61883-1) helper functions
  *
@@ -48,7 +56,12 @@
 		      int (*check)(struct cmp_connection *c, __be32 pcr),
 		      enum bus_reset_handling bus_reset_handling)
 {
-	__be32 old_arg, buffer[2];
+	__be32 old_arg;
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 39)
+	__be32 buffer[2];
+#else
+	__be32 *buffer = c->resources.buffer;
+#endif
 	int err;
 
 	buffer[0] = c->last_pcr_value;
