--- ../alsa-kernel/pci/fm801.c	2013-09-26 10:20:39.774474402 +0200
+++ fm801.c	2013-09-26 16:13:33.707111539 +0200
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /*
  *  The driver for the ForteMedia FM801 based soundcards
  *  Copyright (c) by Jaroslav Kysela <perex@perex.cz>
@@ -1233,7 +1234,7 @@
 		pci_set_master(pci);
 	}
 
-	if (pci->revision >= 0xb1)	/* FM801-AU */
+	if (snd_pci_revision(pci) >= 0xb1)	/* FM801-AU */
 		chip->multichannel = 1;
 
 	snd_fm801_chip_init(chip, 0);
@@ -1445,3 +1446,5 @@
 };
 
 module_pci_driver(fm801_driver);
+
+EXPORT_NO_SYMBOLS;
