--- ../../alsa-kernel/pci/hda/hda_beep.c	2015-03-23 13:20:24.772960667 +0100
+++ hda_beep.c	2015-03-23 13:42:24.569872096 +0100
@@ -1,3 +1,5 @@
+#define __NO_VERSION__
+#include "adriver.h"
 /*
  * Digital Beep Input Interface for HD-audio codec
  *
@@ -109,7 +111,11 @@
 static int snd_hda_beep_event(struct input_dev *dev, unsigned int type,
 				unsigned int code, int hz)
 {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 22)
+	struct hda_beep *beep = dev->private;
+#else
 	struct hda_beep *beep = input_get_drvdata(dev);
+#endif
 
 	switch (code) {
 	case SND_BELL:
@@ -133,7 +139,9 @@
 
 static void turn_off_beep(struct hda_beep *beep)
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 22)
 	cancel_work_sync(&beep->beep_work);
+#endif
 	if (beep->playing) {
 		/* turn off beep */
 		generate_tone(beep, 0);
@@ -146,6 +154,9 @@
 		input_unregister_device(beep->dev);
 	else
 		input_free_device(beep->dev);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 15)
+	kfree(beep->dev);
+#endif
 	beep->dev = NULL;
 	turn_off_beep(beep);
 }
@@ -155,7 +166,11 @@
 	struct input_dev *input_dev;
 	struct hda_codec *codec = beep->codec;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 15)
 	input_dev = input_allocate_device();
+#else
+	input_dev = kzalloc(sizeof(*input_dev), GFP_KERNEL);
+#endif
 	if (!input_dev)
 		return -ENOMEM;
 
@@ -172,7 +187,11 @@
 	input_dev->evbit[0] = BIT_MASK(EV_SND);
 	input_dev->sndbit[0] = BIT_MASK(SND_BELL) | BIT_MASK(SND_TONE);
 	input_dev->event = snd_hda_beep_event;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 22)
 	input_set_drvdata(input_dev, beep);
+#else
+	input_dev->private = beep;
+#endif /* 2.6.22 */ 
 
 	beep->dev = input_dev;
 	return 0;
@@ -278,6 +297,7 @@
 	if (!beep || !beep->dev)
 		return 0;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 15)
 	err = input_register_device(beep->dev);
 	if (err < 0) {
 		codec_err(codec, "hda_beep: unable to register input device\n");
@@ -286,6 +306,9 @@
 		kfree(beep);
 		return err;
 	}
+#else
+	input_register_device(beep->dev);
+#endif
 	beep->registered = true;
 	return 0;
 }
