--- ../../alsa-kernel/drivers/pcsp/pcsp.c	2014-02-14 10:29:56.461067921 +0100
+++ pcsp.c	2014-02-14 10:46:42.548858883 +0100
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /*
  * PC-Speaker driver for Linux
  *
@@ -103,6 +104,9 @@
 		return -EINVAL;
 
 	hrtimer_init(&pcsp_chip.timer, CLOCK_MONOTONIC, HRTIMER_MODE_REL);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 29)
+	pcsp_chip.timer.cb_mode = HRTIMER_CB_IRQSAFE_UNLOCKED;
+#endif
 	pcsp_chip.timer.function = pcsp_do_timer;
 
 	err = snd_card_new(dev, index, id, THIS_MODULE, 0, &card);
