--- ../../alsa-kernel/drivers/pcsp/pcsp.c	2015-07-24 20:09:03.433652027 +0200
+++ pcsp.c	2015-07-27 11:30:20.121244088 +0200
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /*
  * PC-Speaker driver for Linux
  *
@@ -42,8 +43,17 @@
 static int snd_pcsp_create(struct snd_card *card)
 {
 	static struct snd_device_ops ops = { };
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 2, 0)
 	unsigned int resolution = hrtimer_resolution;
+#else
+	unsigned int resolution;
+#endif
 	int err, div, min_div, order;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 2, 0)
+	struct timespec tp;
+	hrtimer_get_res(CLOCK_MONOTONIC, &tp);
+	resolution = tp.tv_nsec;
+#endif
 
 	if (!nopcm) {
 		if (resolution > PCSP_MAX_PERIOD_NS) {
@@ -100,6 +110,9 @@
 		return -EINVAL;
 
 	hrtimer_init(&pcsp_chip.timer, CLOCK_MONOTONIC, HRTIMER_MODE_REL);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 29)
+	pcsp_chip.timer.cb_mode = HRTIMER_CB_IRQSAFE_UNLOCKED;
+#endif
 	pcsp_chip.timer.function = pcsp_do_timer;
 
 	err = snd_card_new(dev, index, id, THIS_MODULE, 0, &card);
