--- ../../alsa-kernel/pci/au88x0/au88x0.c	2014-10-20 08:00:46.676007856 +0200
+++ au88x0.c	2014-10-22 12:21:41.021882073 +0200
@@ -14,7 +14,7 @@
  * Now at least i can play Legacy DOOM with MIDI music :-)
  */
 
-#include "au88x0.h"
+#include "../../alsa-kernel/pci/au88x0/au88x0.h"
 #include <linux/init.h>
 #include <linux/pci.h>
 #include <linux/slab.h>
@@ -338,7 +338,7 @@
 		snd_card_free(card);
 		return err;
 	}
-	chip->rev = pci->revision;
+	chip->rev = snd_pci_revision(pci);
 #ifdef CHIP_AU8830
 	if ((chip->rev) != 0xfe && (chip->rev) != 0xfa) {
 		dev_alert(card->dev,
