--- ../../alsa-kernel/pci/echoaudio/echoaudio.c	2015-01-11 11:56:56.520391385 +0100
+++ echoaudio.c	2015-01-14 15:50:56.609983765 +0100
@@ -1901,6 +1901,7 @@
 	struct echoaudio *chip;
 	int err;
 	size_t sz;
+	u16 subsystem_device;
 	static struct snd_device_ops ops = {
 		.dev_free = snd_echo_dev_free,
 	};
@@ -1972,7 +1973,12 @@
 	chip->comm_page_phys = chip->commpage_dma_buf.addr;
 	chip->comm_page = (struct comm_page *)chip->commpage_dma_buf.area;
 
-	err = init_hw(chip, chip->pci->device, chip->pci->subsystem_device);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 3, 0)
+	subsystem_device = pci->subsystem_device;
+#else
+	pci_read_config_word(pci, PCI_SUBSYSTEM_ID, &subsystem_device);
+#endif
+	err = init_hw(chip, chip->pci->device, subsystem_device);
 	if (err >= 0)
 		err = set_mixer_defaults(chip);
 	if (err < 0) {
@@ -2195,6 +2201,7 @@
 	struct comm_page *commpage, *commpage_bak;
 	u32 pipe_alloc_mask;
 	int err;
+	u16 subsystem_device;
 
 	commpage_bak = kmalloc(sizeof(struct echoaudio), GFP_KERNEL);
 	if (commpage_bak == NULL)
@@ -2202,7 +2209,12 @@
 	commpage = chip->comm_page;
 	memcpy(commpage_bak, commpage, sizeof(struct comm_page));
 
-	err = init_hw(chip, chip->pci->device, chip->pci->subsystem_device);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 3, 0)
+	subsystem_device = pci->subsystem_device;
+#else
+	pci_read_config_word(pci, PCI_SUBSYSTEM_ID, &subsystem_device);
+#endif
+	err = init_hw(chip, chip->pci->device, subsystem_device);
 	if (err < 0) {
 		kfree(commpage_bak);
 		dev_err(dev, "resume init_hw err=%d\n", err);
