--- ../../alsa-kernel/pci/echoaudio/echoaudio.c	2014-11-05 15:38:31.147011646 +0100
+++ echoaudio.c	2014-11-05 15:43:23.744779248 +0100
@@ -1906,6 +1906,7 @@
 	struct echoaudio *chip;
 	int err;
 	size_t sz;
+	u16 subsystem_device;
 	static struct snd_device_ops ops = {
 		.dev_free = snd_echo_dev_free,
 	};
@@ -1977,7 +1978,12 @@
 	chip->comm_page_phys = chip->commpage_dma_buf.addr;
 	chip->comm_page = (struct comm_page *)chip->commpage_dma_buf.area;
 
-	err = init_hw(chip, chip->pci->device, chip->pci->subsystem_device);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 3, 0)
+	subsystem_device = pci->subsystem_device;
+#else
+	pci_read_config_word(pci, PCI_SUBSYSTEM_ID, &subsystem_device);
+#endif
+	err = init_hw(chip, chip->pci->device, subsystem_device);
 	if (err >= 0)
 		err = set_mixer_defaults(chip);
 	if (err < 0) {
@@ -2204,6 +2210,7 @@
 	struct comm_page *commpage, *commpage_bak;
 	u32 pipe_alloc_mask;
 	int err;
+	u16 subsystem_device;
 
 	pci_restore_state(pci);
 	commpage_bak = kmalloc(sizeof(struct echoaudio), GFP_KERNEL);
@@ -2212,7 +2219,12 @@
 	commpage = chip->comm_page;
 	memcpy(commpage_bak, commpage, sizeof(struct comm_page));
 
-	err = init_hw(chip, chip->pci->device, chip->pci->subsystem_device);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 3, 0)
+	subsystem_device = pci->subsystem_device;
+#else
+	pci_read_config_word(pci, PCI_SUBSYSTEM_ID, &subsystem_device);
+#endif
+	err = init_hw(chip, chip->pci->device, subsystem_device);
 	if (err < 0) {
 		kfree(commpage_bak);
 		dev_err(dev, "resume init_hw err=%d\n", err);
